src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -257,7 +238,92 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
                 slice: None,
                 suffix: Vec::new(),
             },
-            _ => PatKind::Constant { value: cv },
+            // If the type is not structurally comparable, just emit the constant directly, causing
+            // the pattern match code to treat it opaquely.
+            _ if !self.type_marked_structural(cv.ty) => PatKind::Constant { value: cv },
+            // Everything below is structurally equal and gets destructured in order to get full
+            // exhaustiveness checks.
+            ty::Ref(_, pointee_ty, ..) if pointee_ty.is_sized(tcx.at(self.span), param_env) => {
+                let old = self.backcompat_hack.replace(true);
+                let val = PatKind::Deref {
+                    subpattern: self.recur(tcx.deref_const(self.param_env.and(cv))),
```

Comment:
What if we did this for unsized pointees too? I don't think we'd have much of an issue with unsized `ty::Const`s.

Reply:
our exhaustiveness and unreachable arm checks can't handle this for `str` and `[u8]` right now. I'd rather not make changes to it while modifying `const_to_pat` in the same PR.